SmartDeviceLink/SDLH264VideoEncoder.m

```diff
@@ -159,6 +159,11 @@ - (BOOL)encodeFrame:(CVImageBufferRef)imageBuffer presentationTimestamp:(CMTime)
 
     OSStatus status = VTCompressionSessionEncodeFrame(_compressionSession, imageBuffer, presentationTimestamp, kCMTimeInvalid, NULL, (__bridge void *)self, NULL);
 
+    if (status == kVTInvalidSessionErr) {
+        VTCompressionSessionCompleteFrames(_compressionSession, presentationTimestamp);
```

Comment:
You can use `kCMTimeInvalid` instead of the timestamp. That way you can move `VTCompressionSessionCompleteFrames` to right before the compression session is invalidated (here and an in the `stop` method - maybe create a new method called `invalidateCompressionSession` since it is duplicate code) .

Reply:
> That way you can move VTCompressionSessionCompleteFrames to right before the compression session is invalidated

Could you clarify what you mean by this? Do you mean to add the `invalidateCompressionSession` method call in the `stop` method and `sdl_resetCompressionSession` method? For example:
```
- (void)sdl_invalidateCompressionSession {
    VTCompressionSessionCompleteFrames(self.compressionSession, kCMTimeInvalid);
    VTCompressionSessionInvalidate(self.compressionSession);
}
```
This would be called in `stop` and also in `sdl_resetCompressionSession` where we are already invalidating the session:
```- (BOOL)sdl_resetCompressionSession {
    // Destroy the current compression session before attempting to create a new one. Otherwise the attempt to create a new compression session sometimes fails.
    if (self.compressionSession != NULL) {
        [self sdl_invalidateCompressionSession];
        CFRelease(self.compressionSession);
        self.compressionSession = NULL;
    }

    OSStatus status = VTCompressionSessionCreate(NULL, (int32_t)self.dimensions.width, (int32_t)self.dimensions.height, kCMVideoCodecType_H264, NULL, self.sdl_pixelBufferOptions, NULL, &sdl_videoEncoderOutputCallback, (__bridge void *)self, &_compressionSession);
    return (status == noErr);
}
```

Reply:
Sorry, I was talking about consolidating the code that destroys the the compression session like so:

```
// TODO: add documentation for this method
- (void)sdl_invalidateCompressionSession {
    if (self.compressionSession == NULL) { return; }

    VTCompressionSessionCompleteFrames(self.compressionSession, kCMTimeInvalid);
    VTCompressionSessionInvalidate(self.compressionSession);
    CFRelease(self.compressionSession);
    self.compressionSession = NULL;
}
```
which would be called by the `stop` and `sdl_resetCompressionSession` methods like so:

```
- (void)stop {
    _currentFrameNumber = 0;
    _timestampOffset = 0.0;

    [self sdl_invalidateCompressionSession];
}
```

```
/// Attempts to create a new VTCompressionSession using the dimensions passed when the video encoder was created and returns whether or not creating the new compression session was created successfully.
- (BOOL)sdl_resetCompressionSession {
    // Destroy the current compression session before attempting to create a new one. Otherwise the attempt to create a new compression session sometimes fails.
    [self sdl_invalidateCompressionSession];

    OSStatus status = VTCompressionSessionCreate(NULL, (int32_t)self.dimensions.width, (int32_t)self.dimensions.height, kCMVideoCodecType_H264, NULL, self.sdl_pixelBufferOptions, NULL, &sdl_videoEncoderOutputCallback, (__bridge void *)self, &_compressionSession);
    return (status == noErr);
}
```